LL-6521 - Swap providers fetching + pairs drilling

imqdee opened the pull request:
<!-- Description of what the PR does go here... screenshot might be good if appropriate -->
Includes the logic to fetch the provider list, constructs pair list and connect selectors from the form

### Type

Feature

<!-- e.g. Bug Fix, Feature, Code Quality Improvement, UI Polish... -->

### Context

<!-- e.g. GitHub issue #45 / contextual discussion -->

Includes #13 
[LL-6516](https://ledgerhq.atlassian.net/secure/RapidBoard.jspa?rapidView=277&modal=detail&selectedIssue=LL-6516&search=6516&assignee=60dc61317d016900702b1806)


### Parts of the app affected / Test plan

<!-- Which part of the app is affected? What to do to test it, any special thing to do? -->

imqdee committed:
chore(swap): create navbar component

imqdee committed:
chore(swap):integrate routing logic based on path

imqdee committed:
design(swap) includes v1 of the history screen

Copy the files from the v1 history directory to the v2 directory and makes
minor ui adjustments to adapt to the new layout.
Didn't touch anything about the logic.

imqdee committed:
i18n(swap): remove unused bullets translation

As asked [here](https://github.com/juan-cortes/ledger-live-desktop/pull/11/files/893e65e456abb967c8e419a90607908061295f10#r676534339)

imqdee committed:
fix(swap/i18n): move from v1 translation to v2

imqdee committed:
feature(swap): create useSwapProviders hooks

This hooks is in charge to fetch providers and filter them based on the
blocklist passed as an env.

imqdee committed:
chore(swap):create swap store & manage form states

This store will be used to store providers data.
- Created actions to store providers and reset state
- Created selectors to fetch provider list and the current provider
- Manage form states from the swapForm component

imqdee committed:
design(swap): integrate form loading screen

imqdee committed:
design(swap): integrate not available form screen

imqdee committed:
chore(swap): move from AvailableProvider v1 to v3

imqdee committed:
chore(swap): enable experimental flag

juan-cortes commented on src/renderer/Default.js:
```diff
@@ -56,7 +56,7 @@ export const TopBannerContainer: ThemedComponent<{}> = styled.div`
   const location = useLocation();
   const ref: React$ElementRef<any> = useRef();
-  const isSwapV2Enabled = true; // useEnv("EXPERIMENTAL_SWAP");
+  const isSwapV2Enabled = useEnv("EXPERIMENTAL_SWAP");
```
This means we need to be yalc'ing live-common or is the version that added the flag already merged?

imqdee replied:
Yep, working like a charm without needing to yalc anything.

juan-cortes replied:
nice

juan-cortes commented on src/renderer/createStore.js:
```diff
@@ -21,8 +21,11 @@ export default ({ state, dbMiddleware }: Props) => {
 
-  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
-  const enhancers = composeEnhancers(applyMiddleware(...middlewares));
+  // const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
```
Why? and if not needed, why not drop it

imqdee replied:
That's an artefact. In fact, I tried to improve the redux configuration to fix the issue with the extension but I didn't test it enough to be sure that the changes fix something. I didn't want to push this change, it should stay stashed from my laptop. I'm going to remove this change.

juan-cortes replied:
OK

imqdee replied:
Removed from [0266e4e93ff6c57c7ef83b0dd1e25c0d9b44aa80](https://github.com/juan-cortes/ledger-live-desktop/pull/14/commits/0266e4e93ff6c57c7ef83b0dd1e25c0d9b44aa80)

juan-cortes commented on src/renderer/reducers/index.js:
```diff
@@ -23,6 +25,7 @@ export type State = {
   settings: SettingsState,
   onboarding: OnboardingState,
   UI: UIState,
+  swap: SwapStateType,
```
So this swap state is not stored on file, correct?
I say this because if it is, then we might as well just bring all the swap namespace from the settings state here too 🤔

imqdee replied:
Yes, it is not stored on file.
The data is only stored in the top-level app's state, if you reboot the app or refresh it, the data stored here is no longer stored.

juan-cortes replied:
Then it makes sense to keep it separate from the `settings.swap` namespace, gotcha

juan-cortes commented on src/renderer/screens/exchange/Swap2/Form/FormLoading/FormLoading.js:
```diff
@@ -0,0 +1,13 @@
+// @flow
```
Why is this not just in index.js? is it a standard that I'm missing?

imqdee replied:
That's a convention that I use to work with, treat components like modules. I use to create a directory to encapsulate components related files and expose only what is supposed to be exposed from the index file which is the entry point of the "module".
For example, suppose that you have a component that is an input that manipulates dates and times, I use to create structure like this.

```sh
InputDateTime/
  InputDateTime.js # The main component
  InputDateTime.test.js # Component's unit tests
  InputDateTime.snap.test.js # Component's snapshot
  InputDateTime.stories.js # Component's story
  index.js #only export main component. If one of the utils files are needed elsewhere, we've to move it to the common utils/ directory
  utils/
    getCurrentDate.js # utils function
    getCurrentDate.test.js # getCurrentDate's unit tests
    getCurrentTime.js # utils function
    getCurrentTime.test.js # getCurrentDate's unit tests
    useNow.js # custom hooks
    useNow.test.js # custom hooks tests
```

That avoid zombies files because if you move for example tests files to a dedicated directory and the same thing for stories when you delete or update the component you've to think about checking other parts of the project structure to remove/update related files. If everything is encapsulated and located in the same directory, you have got a better overview of the components and that will be easy to maintain. Plus, it is easier to localise the lack of tests for example.

--- 

I used that convention because I used to and I find this one powerful. However, I'm not attached to this one and I can move the logic into the index file if you want. 
However, as the team is growing maybe it could be interesting to discuss some conventions in the future, would be cool to have the point of view of everybody and mix them to find the perfect way to structure our projects!

juan-cortes replied:
I like your reasoning, so keep it as you do and we can discuss with the team for a standard that we all like later on.

juan-cortes commented on src/renderer/screens/exchange/Swap2/Form/FormSelectors/FromRow.js:
```diff
@@ -13,11 +13,11 @@ import Text from "~/renderer/components/Text";
 type Props = {
-  fromAccount: ?Account,
-  setFromAccount: (?Account) => void,
+  fromAccount: ?AccountLike,
```
You might want to deal with `account` and `parentAccount` 🤷🏼‍♂️ not a requirement though

imqdee replied:
AccountLike is either `Account`, `TokenAccount` or `ChildAccount`.

```ts
// from *src/types/account.js*
export type SubAccount = TokenAccount | ChildAccount;
export type AccountLike = Account | SubAccount;
```

In another comment, you tell me to don't worry about ChildAccount so perhaps I can just use `Account` and `TokenAccount`?

```ts
type SwappableAccount = Account | TokenAccount;
```

Didn't find any types related to the parent accounts. Are `Account` and `TokenAccount` susceptible to be flagged as parents if they have got some child accounts?

juan-cortes replied:
Parents are always top-level `Account`, so you will have something like this for a swap. 
```
const fromAccount: Account | TokenAccount;
const fromParentAccount?: Account;
```

imqdee replied:
Thanks!

juan-cortes commented on src/renderer/screens/exchange/Swap2/Form/FormSelectors/ToRow.js:
```diff
@@ -46,8 +68,7 @@ export default function ToRow({ toCurrency, setToCurrency, toAmount, setToAmount
             containerProps={amountInputContainerProps}
-            // $FlowFixMe
-            unit={unit}
+            defaultUnit={toCurrency?.units[0]}
```
Better use the `getAccountUnit` helper, in case things change

imqdee replied:
Got it, I'm going to

imqdee replied:
As I'm dealing with currencies here (and not accounts), are we agree that when the currency is set by the user from the select component, I have to derivate the related account from the selected currency and save it to the form state?

Once the account is saved, I'll be able to:
- Dealing with units here as I have the account object associated with the currency
- Dynamically render target account row (the last one) from the form based on the account object

Do I understand correctly?

juan-cortes replied:
That's it, to me, the selection of the `toCurrency` merely does an implicit `toAccount` selection.
So, when I select... BTC I would find the first BTC account as the `toAccount`, and then allow the user to change from that account to another BTC account. If I change the `toCurrency` to ETH, I would find the first ETH account and set it as the `toAccount`

The only added complication is if we select a token currency, I would need to set the `toAccount` and the `toParentAccount` (whatever names you choose, but that's the gist of it)

imqdee replied:
Done.

juan-cortes commented on src/renderer/screens/exchange/Swap2/Form/FormSummary/SectionTarget.js:
```diff
@@ -36,6 +42,31 @@ const TextWrappper: ThemedComponent<{}> = styled.div`
+    <SummarySection>
+      <SummaryLabel label={t("swap2.form.details.label.target")} />
+      <SummaryValue value={targetName} handleChange={() => {}}>
+        {getTargetIcon(targetCurrency)}
```
Why not simply 
```
{targetCurrency? <CryptoCurrencyIcon circle currency={targetCurrency} size={16} /> : null}
```

imqdee replied:
To improve clarity I use to limit computation/conditional from the JSX. From my mind, JSX must be simple to read and understand, avoiding complex map computations or complex conditionals is a good practice.

In this case, it's completely unrelated and I agree with you that I prefer a ternary than calling a function here. I'm going to update the code.

juan-cortes replied:
If there's heavy lifting for sure, but in this particular case I think it's simple enough

imqdee replied:
Apply change there [0293d1eb832852b8c0e977a1cb975bcba69688ab](https://github.com/juan-cortes/ledger-live-desktop/pull/14/commits/0293d1eb832852b8c0e977a1cb975bcba69688ab)

juan-cortes commented on src/renderer/screens/exchange/Swap2/Form/index.js:
```diff
@@ -32,19 +39,37 @@ const SwapForm = () => {
+    );
+
+  // TODO: ensure that the error is catch by Sentry in this case
+  if (error) return <FormNotAvailable />;
```
use the Track component to track the event, and unify those into one if they are the same

imqdee replied:
Do we wanna track page states (loading/error)? 🤔 
Navbar component already sends analytics based on the route, using the `TrackPage` component. 

```js
// from src/renderer/screens/exchange/Swap2/Navbar/index.js
<Nav>
      <TrackPage category="swap" name={swapRoutes[currentIndex].name} />
      <TabBar tabs={tabs} onIndexChange={onWrappedTabChange} index={currentIndex} />
    </Nav>
``` 
However it only sends routing information but nothing related to the state of the page is send (the page is loaded, the page displays the error state).
Do we want this behaviour for metrics usage? If yes, do we have a way to send events programmatically to add more granularity to the last navigation event sent or should I considered loading/error states as "sub-pages" and add a particular suffix to the form page name? For example:

```js
<TrackPage category="swap" name={"history"} /> // from history page
<TrackPage category="swap" name={"form"} /> // from form page
<TrackPage category="swap" name={"form:loading"} /> // from form page when the form is loading
<TrackPage category="swap" name={"form:error"} /> // from form page when the form is displaying the error state
```

juan-cortes replied:
`TrackPage` for screen loads, and the `Track` for events inside of them. 
Track also takes a few props to determine when they get triggered, so they can only load on render, or on unmount, or whatever, maybe they match your needs. I'd say go with something that resembles the events we already have, and then we will need to sync with product on what exactly they want tracked

imqdee replied:
Perfect, I added `Track` component using the same event name that you used from the last version of the swap.

juan-cortes commented on src/renderer/screens/exchange/Swap2/History/HistoryLoading.js:
```diff
@@ -0,0 +1,12 @@
+
+const HistoryLoading = () => (
+  <Box flex={1} alignItems="center" justifyContent="center">
+    <BigSpinner size={50} />
```
I think the loading spinner for providers is larger than this? they should be the same, whatever figma says

imqdee replied:
Got it!

imqdee replied:
Change here [3d35d310bd344ac1099ff4e873ca696371e0ad5c](https://github.com/juan-cortes/ledger-live-desktop/pull/14/commits/3d35d310bd344ac1099ff4e873ca696371e0ad5c)

juan-cortes commented on src/renderer/screens/exchange/Swap2/utils/shared/getAccountLikeId.js:
```diff
@@ -0,0 +1,14 @@
+import type { AccountLike } from "@ledgerhq/live-common/lib/types";
+
+const getAccountId = (fromAccount: ?AccountLike) => {
+  if (!fromAccount) return "";
```
I doubt you need this helper, you have the `getAccountCurrency` helper which will give you the token or currency already, and you can get the id from there directly. Don't worry about `ChildAccount` because they will not be swappable

imqdee replied:
Oh right, thanks, I'll give a try to `getAccountCurrency`

imqdee replied:
Changes here [fa0a5290a1de93c9a89a4577b6b6884412361341](https://github.com/juan-cortes/ledger-live-desktop/pull/14/commits/fa0a5290a1de93c9a89a4577b6b6884412361341)

imqdee committed:
feat(swap):connect providers pairs to form selects

From now, values from the **to** selector is based on the available pairs
and the value of the **from** selector.
I take the opportunity to rework the way state will be drilled to the
details session.

imqdee committed:
review: remove getAccountId custom helper

imqdee committed:
review: rewrite target account to a ternary

imqdee committed:
review: increase history loading spinner size

imqdee committed:
review: add track events

imqdee committed:
review: deal with account instead of tokens from ToRow

Ready to move to the next step, I now have everything to leave the selectors
and implement logic to handle all other features from the same section (
max feature, invert feature, and handle amount changes)

imqdee committed:
review: use getAccountUnit helper

imqdee commented:
👋  @juan-cortes 
Added some changes related to your comments and I worked on the *toSelect* to deal with account instead of currency directly.
I now have a clear vision of what is missing and I have the feeling that I did the hardest part. The next steps will be
- Connect amount input to allow users to type amounts and save the value if allowed
- Move selectors state from one-level upper and transform it as a reducer
- Develop invert and max feature
- Connect the details section to the state to fetch data from the state
- Develop interaction from the details section (rate selector etc...)
- CLEAN. This step will be important, my vision is more and more clear, I know that I'll have to rework some parts and remove some others.

Here is a screen recording of the selectors and the temporary state.
[Drive link](https://drive.google.com/file/d/15F4NHSWnWfNnag0K4Ak-hc0p3CjGZTz-/view?usp=sharing)

I suppose that for the moment, my code looks drafty so if you want, we talk discuss about it and share our vision about the next steps and how to handle them.

juan-cortes commented:
The video looks neat, I'm going to review the latest version and hopefully, we can merge this back into the main PR before we move onto the rebranding for the rest of the week

juan-cortes commented on src/renderer/components/PerCurrencySelectAccount/state.js:
```diff
@@ -46,6 +46,14 @@ const getIdsFromTuple = (accountTuple: AccountTuple) => ({
+export type useCurrencyAccountSelectReturnType = {
+  availableAccounts: Array<AccountTuple>,
+  currency: ?CryptoCurrency | TokenCurrency,
+  account: ?Account | any,
```
I prefer we use the account / parentAccount naming convention since we use it throughout the whole codebase.
And `SubAccount` might give you problems since it will try to allow for `ChildAccount` too, that's why we were using `TokenAccount | Account` elsewhere, if you can make it work like that it's fine by me, but account/parentAccount 🙏🏼

imqdee replied:
Oh right, I'm going to rework that hook and replace `SubAccount` with `parentAccount`. Hope other pieces of code that use it (outside of the swap) is covered by tests 👀

imqdee replied:
Done, as we discussed on Slack, I casted types received from the hook in waiting reworking.

juan-cortes commented on src/renderer/components/PerCurrencySelectAccount/state.js:
```diff
@@ -46,6 +46,14 @@ const getIdsFromTuple = (accountTuple: AccountTuple) => ({
+  account: ?Account | any,
+  subAccount: ?SubAccount | any,
+  setAccount: (account: ?Account, subAccount: ?SubAccount) => void,
+  setCurrency: (currency: ?CryptoOrTokenCurrency) => void,
```
Be explicit here and do `(CryptoCurrency | TokenCurrency)` as we do everywhere else too

imqdee replied:
Done!

juan-cortes commented on src/renderer/screens/exchange/Swap2/Form/FormSelectors/FormInputs.js:
```diff
@@ -20,12 +31,57 @@ function SwapButton() {
+      targetAccountExists: true,
+      data: {
+        account: ?Account,
+        subAccount: ?SubAccount,
```
account/parentAccount

imqdee replied:
Done!

juan-cortes commented on src/renderer/screens/exchange/Swap2/Form/FormSelectors/FormInputs.js:
```diff
@@ -20,12 +31,57 @@ function SwapButton() {
   const [toAmount, setToAmount] = useState(null);
 
+  // TODO: would be handled by the reducer in the future
+  const handleSetToAccountChange = (selectSate: useSelectableCurrenciesReturnType) => {
```
There's a reducer in logic.js in live-common for swap that does just this, perhaps you can rework it, or redo it, or something.

imqdee replied:
Yeah definitely. I plan to use/modify it once I'll start to work on the reducer logic (moving states one-step higher, from FromInput to the form container)

juan-cortes replied:
👍🏼

juan-cortes commented on src/renderer/screens/exchange/Swap2/Form/FormSelectors/ToRow.js:
```diff
@@ -1,40 +1,74 @@
 type Props = {
-  toCurrency: ?CryptoCurrency,
-  setToCurrency: (?CryptoCurrency) => void,
+  fromAccount: ?SwappableAccountType,
```
Did you wrap this type into its own thing for a specific reason? I'm so used to seeing the verbose
`Account | TokenAccount` everywhere that strikes me as odd.

imqdee replied:
I wrapped this type to avoid explicitly call `Account | TokenAccount` everywhere multiple times but as I understand from your comments, this is the convention that the team choose so I will remove my wrapper and replace it.

imqdee replied:
Done!

juan-cortes reviewed (changes requested):
Just a couple of pointers and we can merge into the main one

imqdee committed:
review: being explicit with Account | TokenAccount types

imqdee committed:
review: use account/parentAccount convention

imqdee commented:
@juan-cortes Take all of your messages into account and pushed a new version that implements account/parentAccount as you can see in this record.

[GDrive link](https://drive.google.com/file/d/1gY1QSGUP3vJ8zpAIxSxUh7ViJe896QAz/view?usp=sharing)

During the recording I only logged account/parentAccount keys from the state, not all the state is logged of course (for example, currency is saved but not logger here).
